### PR TITLE
bpf: read and copy proc exe at execve for matchBinaries

### PR DESCRIPTION
**Pull request by mtardy**
Fix for https://github.com/cilium/tetragon/issues/1741

This PR changes how we read the binary path at `execve` time. We now read `exe` from the `task_struct` instead of relying upon the tracepoint argument that was just the path given to `execve` (thus possibly relative). This was an issue since `matchBinaries` could be avoided using relative links instead of absolute ones at `execve`.

### For a future patch maybe:

In addition to this, we could also use this `exe` value for the binary path of the event. We also previously used the argument of `execve`, being possibly relative and corrected in userspace using `cwd`, but also possibly a symlink. This would thus introduce a user-facing change: tetragon events will now contain the canonical binary path while previously it could have been a symlink.

Without that user might try to "matchBinaries" a symlink file displayed in the Tetragon events and fail to make it work.

Preliminary work would be to make `prepend_name` work with 4096 (or `MAX_PATH`) buffer size.

### CODE REVIEW

**Comment by mtardy** [comment]
Unfortunately, it looks like it doesn't fix for 4.19, I'll have to put it out for small progs kernels :/

File: bpf/lib/process.h [hunk 1]
```diff
@@ -270,6 +270,19 @@ struct msg_k8s {
 	char docker_id[DOCKER_ID_LENGTH];
 }; // All fields aligned so no 'packed' attribute.
 
+#define BINARY_PATH_MAX_LEN 256
+
+struct heap_exe {
+	// because of verifier limitations, this has to be 2 * 256 bytes while 256
```
Comment (jrfastab) [inline bpf/lib/process.h:276]: Is this true with large programs? We might be able to widen this some if we feel its an actual issue.

File: bpf/lib/process.h [hunk 2]
```diff
@@ -270,6 +270,19 @@ struct msg_k8s {
...
+	// should be theoretically sufficient, and actually is, in unit tests.
+	char buf[BINARY_PATH_MAX_LEN * 2];
```
Comment (jrfastab) [inline bpf/lib/process.h:278]: If we were a bit smarter this can be improved 2x waste of space. Over 256B though I don't care to do the work.

**Review by jrfastab — APPROVED** [review]

**Comment by jrfastab** [comment]
OK lets try this. Merging I want to land this in a test cluster and play with it though so I'll do that shortly.

**Comment by tixxdz** [comment]
> OK lets try this. Merging I want to land this in a test cluster and play with it though so I'll do that shortly.

So I wanted to suggest for security reasons and be bullet proof go inside https://github.com/cilium/tetragon/blob/main/bpf/process/bpf_execve_bprm_commit_creds.c#L47  using bprm->file

As it is now on the tracepoint hook is **somehow** late the window race is so small and I don't think it can be triggered to be honest, but I prefer no surprise ;-)
The `allow_write_access(exec) ` https://elixir.bootlin.com/linux/latest/source/fs/exec.c#L1789 allows an fput on the file and even if there is some RCU lifetime on some fields of the file struct which are used to resolve the path, let's just ensure for this case that we do it properly under the right locks and avoid dereferencing pointers like that especially if this code is triggered by everyone, not to mention it is core logic of binary handling and we have safe easy alternative.   The other suggestion on bprm->file is totally safe, the bprm is locked.

**Pull request merged**